### PR TITLE
UI crud questions #16

### PR DESCRIPTION
**Pull request by Iammonis**

### CODE REVIEW

File: client/web/src/Routes/Routes.jsx [hunk 1]
```diff
@@ -22,6 +23,9 @@ const Route = () => {
           <PublicRoute path="/login">
             <Login />
           </PublicRoute>
+          <AdminRoute exact path="/api/question" >
+            <Questions />
```
Comment (aravindsakthivel) [inline client/web/src/Routes/Routes.jsx]: Don't add API in the frontend path

File: client/web/src/Structure/Admin/Components/AddQuestion.jsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import { adminActions } from '../State/action'
+
+export const AddQuestion = ({setOpen}) => {
+    const [type, setType] = useState('mcq')
+    const dispatch = useDispatch()
+
+    const handleAdd = () => {
+        let payload = {
+
+        }
```
Comment (aravindsakthivel) [inline client/web/src/Structure/Admin/Components/AddQuestion.jsx:12]: Why payload was an empty object

File: client/web/src/Structure/Admin/Components/AllQuestions.jsx [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+      setPage(0);
+    };
+
+    useEffect(() => {
+        dispatch( adminActions.getQuestionsRequest(page, rowsPerPage))
+    }, [page, rowsPerPage])
+    
+    return (
+           questions.questions !== undefined ? <>
+                <Table>
```
Comment (aravindsakthivel) [inline client/web/src/Structure/Admin/Components/AllQuestions.jsx:34]: Use a isloading to conditionally render, Don't use questions.questions

File: client/web/src/Structure/Admin/Components/QuestionsByTopic.jsx [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import Table from '@material-ui/core/Table';
+import TableHead from '@material-ui/core/TableHead';
+import TableRow from '@material-ui/core/TableRow';
+import TableCell from '@material-ui/core/TableCell';
+import TableBody from '@material-ui/core/TableBody';
+import TablePagination from '@material-ui/core/TablePagination';
+
+export const QuestionsByTopic = ({val, handleDelete}) => {
+    const dispatch = useDispatch();
+    const {data} = useSelector( state => state.admin )
```
Comment (aravindsakthivel) [inline client/web/src/Structure/Admin/Components/QuestionsByTopic.jsx]: don't use destructuring in useSelector, will lead to extra render. dot reference it.

File: client/web/src/Structure/Admin/Components/QuestionsByTopic.jsx [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+      setPage(0);
+    };
+
+    useEffect(() => {
+        dispatch( adminActions.getQuestionsByTopicRequest(val) )
+    }, [])
+
+    return (
+        data && data.questions.length > 0 ? <>
+                <Table>
```
Comment (aravindsakthivel) [inline client/web/src/Structure/Admin/Components/QuestionsByTopic.jsx:34]: Try to use a Isloading if possible

File: client/web/src/Structure/Admin/State/action.js [hunk 6]
```diff
@@ -230,6 +230,165 @@ const updateCrudTopics = (payload, id) => async (dispatch) => {
...
+    type: adminConstants.GET_QUESTIONS_BY_TOPIC_FAILURE,
+    payload: data
+  });
+  
+const getQuestionsByTopicRequest = (topic) => (dispatch) => {
+    console.log(topic)
+    dispatch(getQuestionsByTopicLoading());
+    const token = (localStorage.getItem('token'))
+    let url = `http://localhost:5050/api/question/byTopic/${topic}`
+
```
Comment (aravindsakthivel) [inline client/web/src/Structure/Admin/State/action.js:315]: hide the `API` in env files

File: client/web/src/Structure/Admin/State/action.js [hunk 7]
```diff
@@ -230,6 +230,165 @@ const updateCrudTopics = (payload, id) => async (dispatch) => {
...
+  
+  const addQuestionsFailure = (data) => ({
+    type: adminConstants.ADD_QUESTION_FAILURE,
+    payload: data
+  });
+  
+const addQuestionsRequest = (payload) => (dispatch) => {
+    dispatch(addQuestionsLoading());
+    const token = (localStorage.getItem('token'))
+    let url = `http://localhost:5050/api/question/add/:topic/`
```
Comment (aravindsakthivel) [inline client/web/src/Structure/Admin/State/action.js]: Hide the api. you can check flashkort for it

File: client/web/src/Structure/Admin/State/action.js [hunk 8]
```diff
@@ -230,6 +230,165 @@ const updateCrudTopics = (payload, id) => async (dispatch) => {
...
+  
+  const deleteQuestionsFailure = (data) => ({
+    type: adminConstants.DELETE_QUESTION_FAILURE,
+    payload: data
+  });
+  
+const deleteQuestionsRequest = (id, topic) => (dispatch) => {
+    dispatch(deleteQuestionsLoading());
+    const token = (localStorage.getItem('token'))
+    let url = `http://localhost:5050/api/question/delete/${topic}/${id}`
```
Comment (aravindsakthivel) [inline client/web/src/Structure/Admin/State/action.js]: Hide the API

File: client/web/src/Structure/Admin/State/action.js [hunk 9]
```diff
@@ -230,6 +230,165 @@ const updateCrudTopics = (payload, id) => async (dispatch) => {
...
+    axios({
+      method: "post",
+      url: url,
+      headers: {
+        "Authorization": `Bearer ${token}`,
+        "Content-Type": "application/json",
+      }
+    })
+      .then((res) => { console.log(res.data); return dispatch(deleteQuestionsSuccess(res.data)) } )
+      .catch((err) => dispatch(deleteQuestionsFailure(err)));
```
Comment (aravindsakthivel) [inline client/web/src/Structure/Admin/State/action.js:390]: remove unnecessary console.log(), use the inbuilt return of arrow function

File: client/web/src/Structure/Admin/State/action.js [hunk 10]
```diff
@@ -250,5 +409,10 @@ export const adminActions = {
     updateCrudTopicsRequest,
     updateCrudTopicsSuccess,
     updateCrudTopicsFailure,
-    updateCrudTopics
+    updateCrudTopics,
+    getQuestionsRequest,
+    getTopicsRequest,
+    getQuestionsByTopicRequest,
+    deleteQuestionsRequest,
+    addQuestionsRequest
```
Comment (aravindsakthivel) [inline client/web/src/Structure/Admin/State/action.js:403]: export only required function, the failure, success, request. it won't be used anywhere, remove unnecessay from object

File: client/web/src/Structure/Authentication/state/action.js [hunk 11]
```diff
@@ -43,6 +43,7 @@ const userLoginProcess = () => async (dispatch) => {
 
   try {
     const res = await axios(config);
+    console.log(res.data)
     const { user, token } = res.data
```
Comment (aravindsakthivel) [inline client/web/src/Structure/Authentication/state/action.js:47]: remove console.log

**Review by aravindsakthivel — CHANGES_REQUESTED** [review]
Make the necessary changes.

File: client/web/src/Structure/Admin/State/action.js (hunk 10 above)
Comment (Iammonis) [reply client/web/src/Structure/Admin/State/action.js:403]: This is actually done by @Anantjain226

File: client/web/src/Structure/Admin/Components/AddQuestion.jsx (hunk 2 above)
Comment (Iammonis) [reply client/web/src/Structure/Admin/Components/AddQuestion.jsx:12]: Still working on it

File: client/web/src/Routes/Routes.jsx (hunk 1 above)
Comment (aravindsakthivel) [reply client/web/src/Routes/Routes.jsx]: There is already a route called questions for the private route, so pull the latest and make changes to admin route question

File: client/web/src/Structure/Admin/State/action.js [hunk 12]
```diff
@@ -230,25 +231,173 @@ const updateCrudTopics = (payload, id) => async (dispatch) => {
...
+  const getQuestionsFailure = (data) => ({
+    type: adminConstants.GET_ALL_QUESTIONS_FAILURE,
+    payload: data
+  });
+  
+  const getQuestionsRequest = (page = 1, limit = 10) => (dispatch) => {
+    dispatch(getQuestionsLoading());
+    const token = (localStorage.getItem('token'))
+    let url = `${QUESTION_URL}/all/?page=${page+1}&limit=${limit}`
+
```
Comment (aravindsakthivel) [inline client/web/src/Structure/Admin/State/action.js:252]: default value of page is one and you are increasing it by 1 so you will get second page. so check this

File: client/web/src/Structure/Admin/State/action.js [hunk 13]
```diff
@@ -2,6 +2,7 @@ import {adminConstants} from './actionTypes'
 import axios from 'axios'
 
 const TOPIC_API = process.env.REACT_APP_ADMIN_TOPIC_API_URL
+const QUESTION_URL = process.env.REACT_APP_QUESTION_URL;
```
Comment (aravindsakthivel) [inline client/web/src/Structure/Admin/State/action.js:6]: Give name as `REACT_APP_ADMIN_QUESTION_API_URL` instead as `REACT_APP_QUESTION_URL`

**Review by aravindsakthivel — CHANGES_REQUESTED** [review]
First `PULL` the latest and make changes,

**Review by aravindsakthivel — APPROVED** [review]
Everything is fine

**Pull request merged**